bootstrap/src/lib/cache/rateLimit.ts

```diff
@@ -0,0 +1,140 @@
+  getParticipantMaxAge: (participantId: string) => number | false
+  incrementParticipantHeartbeat: (participantId: string, cost?: number) => number | false
+}
+
+type Participant = {
+  cost: number
+  heartbeat: number
+  lastSeen: number
+}
+
+const totalHeartbeat = new Map()
+let participants: Map<string, Participant> = new Map()
+
+export const makeRateLimit = (options: RateLimitOptions): RateLimit => {
+  const minHeartbeat = 0
+  const safeCapacity = options.totalCapacity * 0.9
+
+  const _isEnabled = (): boolean => {
+    return !!options.groupId && !!options.totalCapacity
+  }
+
+  // Weight normalizes the relevance the participant is having in the adapter
+  const _getWeight = (participantHeartbeat: number, totalHeartbeat: number): number => {
+    // Avoid having Infinity weight
+    if (participantHeartbeat === 0) {
+      participantHeartbeat = 1
+    }
+    if (totalHeartbeat === 0) {
+      totalHeartbeat = 1
+    }
+    return (participantHeartbeat - minHeartbeat) / (totalHeartbeat - minHeartbeat)
+  }
+
+  const _getParticipant = (participantId: string): Participant => {
+    return (
+      participants.get(participantId) || { cost: 1, heartbeat: 0, lastSeen: new Date().getTime() }
+    )
+  }
+
+  const _getParticipantEvenCapacity = (participantId: string) => {
+    const totalParticipants = participants.size || 1
+    return safeCapacity / totalParticipants / _getParticipant(participantId).cost
+  }
+
+  const _getParticipantMaxAge = (participantId: string): number => {
+    const SEC_IN_MIN = 60
+    const MS_IN_SEC = 1000
+
+    // Precise Capacity
+    // const participantHeartbeat = _getParticipant(participantId).heartbeat
+    // const totalHeartbeat = _getCurrentHeartbeat()
+    // const participantWeight = _getWeight(participantHeartbeat, totalHeartbeat)
+    // const allowedReqPerMin = safeCapacity * participantWeight
+
+    const allowedReqPerMin = _getParticipantEvenCapacity(participantId)
+    console.log('PARTICIPANT CAPACITY: ', allowedReqPerMin)
+
+    let maxAge = Math.round(MS_IN_SEC / (allowedReqPerMin / SEC_IN_MIN))
+    if (maxAge < DEFAULT_CACHE_MAX_AGE) {
+      maxAge = DEFAULT_CACHE_MAX_AGE
+    } else if (maxAge >= MAX_AGE_ALLOWED) {
+      maxAge = MAX_AGE_ALLOWED
+    }
+    return maxAge
+  }
+
+  const _getUpdatedParticipant = (participantId: string, cost?: number): Participant => {
+    const participant = _getParticipant(participantId)
+    return {
+      cost: cost || participant.cost,
+      heartbeat: participant.heartbeat + 1,
+      lastSeen: new Date().getTime(),
+    }
+  }
+
+  // hash(API_KEY):uuid()
+  const _getHeartbeatKey = () => {
+    return `${options.groupId}:${options.id}`
+  }
+
+  const _getCurrentHeartbeat = (): number => {
+    return Number(totalHeartbeat.get(_getHeartbeatKey())) || 0
+  }
+
+  const _incrementTotalHeartbeat = (): number => {
+    const heartbeat = _getCurrentHeartbeat()
+    totalHeartbeat.set(_getHeartbeatKey(), heartbeat + 1)
+    return heartbeat + 1
```

Comment:
Do we need to return anything here? Seems like its used purely for its side-effect